From 16b8662c27926126a431a77f459af8799e2ee172 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Tue, 7 Aug 2018 18:37:06 +0800
Subject: [PATCH 4312/5242] MLK-19152-3 drm/imx: lcdif: add horizontal crop
 support to plane update

commit  c46b377ddec0a460f73cac7e2552ea4dc2da87a1 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add horizontal cropping support when atomic plane update is
running, and if the attached CRTC needs modeset. And if the
width of visible portion of plane is equal to the fb surface
width, the Pigeon Mode will be disabled, so cropping will be
disabled.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit 30672b2b18a07a2926979cc533cbb84ea4a642dd)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/lcdif/lcdif-plane.c |   12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/imx/lcdif/lcdif-plane.c b/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
index 2045289..91bd6fc 100644
--- a/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
+++ b/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
@@ -114,7 +114,8 @@ static void lcdif_plane_atomic_update(struct drm_plane *plane,
 	struct drm_plane_state *state = plane->state;
 	struct drm_framebuffer *fb = state->fb;
 	struct drm_gem_cma_object *gem_obj = NULL;
-	u32 fb_addr, src_off, fb_idx;
+	u32 fb_addr, src_off, src_w, fb_idx;
+	bool crop;
 
 	/* plane and crtc is disabling */
 	if (!fb)
@@ -143,6 +144,15 @@ static void lcdif_plane_atomic_update(struct drm_plane *plane,
 
 	lcdif_set_fb_addr(lcdif, fb_idx, fb_addr);
 
+	/* config horizontal cropping if crtc needs modeset */
+	if (unlikely(drm_atomic_crtc_needs_modeset(state->crtc->state))) {
+		src_w = state->src_w >> 16;
+		WARN_ON(src_w > fb->width);
+
+		crop  = src_w != fb->width ? true : false;
+		lcdif_set_fb_hcrop(lcdif, src_w, fb->width, crop);
+	}
+
 	lcdif_enable_controller(lcdif);
 }
 
-- 
1.7.9.5

